### ISSUE TITLE
Excessive CPU usage, server not responsive

### ISSUE THREAD

**Issue by CherryDT** [input-output-hk/cardano-graphql#499]
### Environment

```
                      _
                     | |
  ____   ____   ____ | |
 / ___) / _  | / _  || |
( (___ ( ( | || | | || |
 \____) \_|| | \_|| ||_|
       (_____|    |_|
{
  os: {
    arch: 'x64',
    cpus: { count: 4, model: 'AMD EPYC 7571' },
    release: '5.4.0-1048-aws',
    totalMem: 16643039232,
    type: 'Linux'
  },
  version: '4.0.0'
}
```

#### Platform

- [X] Linux / Ubuntu
- [ ] Linux / Other
- [ ] macOS
- [ ] Windows

**Platform version**: 

#### Runtime

- [ ] Node.js
- [X] Docker

**Runtime version**: 
```console
$ node -v
v14.17.0
$ docker version
Client:
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.8
 Git commit:        20.10.2-0ubuntu1~20.04.2
 Built:             Tue Mar 30 21:24:57 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.8
  Git commit:       20.10.2-0ubuntu1~20.04.2
  Built:            Mon Mar 29 19:10:09 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.3-0ubuntu2.3
  GitCommit:
 runc:
  Version:          spec: 1.0.2-dev
  GitCommit:
 docker-init:
  Version:          0.19.0
  GitCommit:
```

### Steps to reproduce the bug

I'm not sure, it was running normally, I hadn't run any special queries, just some `tip` and some `utxos()` with a `where { address { _eq: ... } } }`. Last thing I ran before it locked up was a `submitTransaction` mutation.

Then, every query started hanging (for several minutes!) and I saw 100% CPU usage by `postgres`.

`select * from pg_stat_activity;` showed these things happening (it was 08:16 UTC when I ran the command):

```
 datid |  datname  | pid  | usesysid | usename  | application_name | client_addr | client_hostname | client_port |         backend_start         |          xact_start           |          query_start          |         state_change          | wait_event_type |     wait_event      | state  | backend_xid | backend_xmin |                                                                                                                                                                                                                                                                                              query                                                                                                                                                                                                                                                                                              |         backend_type         
-------+-----------+------+----------+----------+------------------+-------------+-----------------+-------------+-------------------------------+-------------------------------+-------------------------------+-------------------------------+-----------------+---------------------+--------+-------------+--------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------
       |           |   25 |          |          |                  |             |                 |             | 2021-05-18 15:40:15.444583+00 |                               |                               |                               | Activity        | AutoVacuumMain      |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | autovacuum launcher
       |           |   27 |       10 | postgres |                  |             |                 |             | 2021-05-18 15:40:15.448334+00 |                               |                               |                               | Activity        | LogicalLauncherMain |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | logical replication launcher
 16384 | cexplorer |   39 |       10 | postgres |                  | 172.18.0.5  |                 |       60200 | 2021-05-18 15:40:16.762821+00 |                               | 2021-05-18 15:40:17.114001+00 | 2021-05-18 15:40:17.114122+00 | Client          | ClientRead          | idle   |             |              | LISTEN  hasura_schema_update;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | client backend
 16384 | cexplorer |   50 |       10 | postgres |                  | 172.18.0.6  |                 |       40332 | 2021-05-18 15:40:16.965624+00 |                               | 2021-05-20 08:18:16.986521+00 | 2021-05-20 08:18:16.988926+00 | Client          | ClientRead          | idle   |             |              | COMMIT                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | client backend
 16384 | cexplorer | 9431 |       10 | postgres |                  |             |                 |             | 2021-05-20 07:49:40.190836+00 | 2021-05-20 07:49:40.187+00    | 2021-05-20 07:49:40.188057+00 | 2021-05-20 07:49:40.192007+00 | IO              | BufFileRead         | active |             |       165770 | SELECT  json_build_object('aggregate', json_build_object('sum', json_build_object('value', sum("value" ) ) ) ) AS "root" FROM  (SELECT  "_0_root.base"."value" AS "value" FROM  (SELECT  *  FROM "public"."Utxo"  WHERE ('true')     ) AS "_0_root.base"      ) AS "_1_root"                                                                                                                                                                                                                                                                                                                    | parallel worker
 16384 | cexplorer | 9430 |       10 | postgres |                  |             |                 |             | 2021-05-20 07:49:40.194066+00 | 2021-05-20 07:49:40.187+00    | 2021-05-20 07:49:40.188057+00 | 2021-05-20 07:49:40.195175+00 | IPC             | MessageQueueSend    | active |             |       165770 | SELECT  json_build_object('aggregate', json_build_object('sum', json_build_object('value', sum("value" ) ) ) ) AS "root" FROM  (SELECT  "_0_root.base"."value" AS "value" FROM  (SELECT  *  FROM "public"."Utxo"  WHERE ('true')     ) AS "_0_root.base"      ) AS "_1_root"                                                                                                                                                                                                                                                                                                                    | parallel worker
 16384 | cexplorer | 9494 |       10 | postgres |                  | 172.18.0.5  |                 |       43528 | 2021-05-20 08:09:15.090312+00 |                               | 2021-05-20 08:18:35.133296+00 | 2021-05-20 08:18:35.13342+00  | Client          | ClientRead          | idle   |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                +| client backend
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |       UPDATE hdb_catalog.event_log                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |       SET locked = 't'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |       WHERE id IN ( SELECT l.id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                     FROM hdb_catalog.event_log l                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                     WHERE l.delivered = 'f' and l.error = 'f' and l.locked = 'f'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                           and (l.next_retry_at is NULL or l.next_retry_at <= now())                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                           and l.archived = 'f'                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                     ORDER BY created_at                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                     LIMIT $1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                     FOR UPDATE SKIP LOCKED )                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |       RETURNING id, schema_name, table_name, trigger_name, payload::json, tries, created_at                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    +| 
       |           |      |          |          |                  |             |                 |             |                               |                               |                               |                               |                 |                     |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | 
 16384 | cexplorer | 9549 |       10 | postgres | psql             |             |                 |          -1 | 2021-05-20 08:18:35.34138+00  | 2021-05-20 08:18:35.351114+00 | 2021-05-20 08:18:35.351114+00 | 2021-05-20 08:18:35.351124+00 |                 |                     | active |             |       165929 | select * from pg_stat_activity;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | client backend
 16384 | cexplorer |   55 |       10 | postgres |                  | 172.18.0.3  |                 |       36700 | 2021-05-18 15:40:18.38704+00  |                               | 2021-05-20 08:18:29.147239+00 | 2021-05-20 08:18:29.14743+00  | Client          | ClientRead          | idle   |             |              | SELECT pg_backend_pid()                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | client backend
 16384 | cexplorer | 9406 |       10 | postgres |                  | 172.18.0.5  |                 |       40170 | 2021-05-20 07:45:24.071702+00 | 2021-05-20 07:49:40.187+00    | 2021-05-20 07:49:40.188057+00 | 2021-05-20 07:49:40.188058+00 | IO              | DataFileRead        | active |             |       165770 | SELECT  json_build_object('aggregate', json_build_object('sum', json_build_object('value', sum("value" ) ) ) ) AS "root" FROM  (SELECT  "_0_root.base"."value" AS "value" FROM  (SELECT  *  FROM "public"."Utxo"  WHERE ('true')     ) AS "_0_root.base"      ) AS "_1_root"                                                                                                                                                                                                                                                                                                                    | client backend
 16384 | cexplorer | 9498 |       10 | postgres |                  | 172.18.0.5  |                 |       43746 | 2021-05-20 08:10:14.738261+00 | 2021-05-20 08:18:33.616735+00 | 2021-05-20 08:18:33.616991+00 | 2021-05-20 08:18:33.616992+00 |                 |                     | active |             |       165929 | SELECT  coalesce(json_agg("root" ORDER BY "root.pg.assetId" ASC NULLS LAST), '[]' ) AS "root" FROM  (SELECT DISTINCT ON ("root.pg.assetId") "_0_root.base"."assetId" AS "root.pg.assetId", row_to_json((SELECT  "_1_e"  FROM  (SELECT  "_0_root.base"."assetId" AS "assetId", "_0_root.base"."assetName" AS "assetName", "_0_root.base"."policyId" AS "policyId"       ) AS "_1_e"      ) ) AS "root" FROM  (SELECT  *  FROM "public"."TokenMint"  WHERE ('true')     ) AS "_0_root.base"    ORDER BY "root.pg.assetId" ASC NULLS LAST LIMIT 500 OFFSET ('157000')::integer) AS "_2_root"       | client backend
       |           |   23 |          |          |                  |             |                 |             | 2021-05-18 15:40:15.444022+00 |                               |                               |                               | Activity        | BgWriterMain        |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | background writer
       |           |   22 |          |          |                  |             |                 |             | 2021-05-18 15:40:15.444536+00 |                               |                               |                               | Activity        | CheckpointerMain    |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | checkpointer
       |           |   24 |          |          |                  |             |                 |             | 2021-05-18 15:40:15.447605+00 |                               |                               |                               | Activity        | WalWriterMain       |        |             |              |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | walwriter
(14 rows)
```

### What is the expected behavior?

Normal operation.

**Comment by rhyslbw** [comment]
Did this issue resolve itself? Have you seen it again? It's the first time I've seen it reported, so any additional info would help.

**Comment by CherryDT** [comment]
After a few minutes it resolved itself, but I don't know yet whether it would come back randomly or not because further development and testing with Cardano is currently blocked by https://github.com/Emurgo/cardano-serialization-lib/issues/156 for me unfortunately.

**Comment by rhyslbw** [comment]
A new version will be released with resource optimisations on background operations, and will use `cardano-db-sync@10`. This version optimises the the write strategy at start of a new epoch, which is probably unrelated, but worth knowing about! I'll keep this issue open until it's released.

**Comment by CherryDT** [comment]
Ok thank you. I ran into the same thing again yesterday so whatever it is, it doesn't seem to be resolved so far 

[Sent from my mobile - please excuse typos]

________________________________

**Comment by ElGatoLoco** [comment]
Is this supposed to be resolved in cardano-db-sync@10+? Our server is getting clogged with dozens of postgres processes, so we were thinking about migrating to newer version of db-sync, but are unsure of its compatibility with cardano-graphql@4 or stability of cardano-graphql@5.
Any update on the issue?

**Comment by svorcan** [comment]
I have the same issue. Do you have any updates on the timeline for the next release?

This is heavily affecting our development and timelines.

**Comment by CyberCyclone** [comment]
For those that are still having issues, what hardware are you running it on? I've been running cardano-graphql:4.0.0 since it's release and it's only had issues on very complex queries, but on 3 servers it's never bogged down for no reason.

I noticed the OP is running on AWS, which from my testing is typically the worst platform to use any form of database on. While I've found having a high-frequency CPU can help, IOPS typically have a greater impact on databases. Unless you're paying hundreds of dollars a month, you'll probably get about 3000 IOPS on AWS, Azure or Google Cloud. So there's a chance that the excessive CPU is actually the CPU waiting on the drive.

Vultr or DigitalOcean typically have the greatest IOPS, from 75,000 (standard) to 150,000 (premium).
